### PR TITLE
8290780: AArch64: Crash in c2 nmethod running RunThese30M.java

### PR DESCRIPTION
**Pull request by theRealAph**
Fix that masks the offsets used when adrp() is passed an unreachable destination. This reloc allows e.g. `adrp; movk; ldr` to access anywhere in the address space.

```
#  SIGSEGV (0xb) at pc=0x0000ffff55964edc, pid=2843096, tid=2850366
#
# JRE version: Java(TM) SE Runtime Environment (20.0+7) (fastdebug build 20-ea+7-377)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 20-ea+7-377, compiled mode, sharing, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# J 91101 c2 java.io.ObjectOutputStream.enableReplaceObject(Z)Z java.base@20-ea (47 bytes) @ 0x0000ffff55964edc [0x0000ffff55964e80+0x000000000000005c]
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290780](https://bugs.openjdk.org/browse/JDK-8290780): AArch64: Crash in c2 nmethod running RunThese30M.java

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9615/head:pull/9615` \
`$ git checkout pull/9615`

Update a local copy of the PR: \
`$ git checkout pull/9615` \
`$ git pull https://git.openjdk.org/jdk pull/9615/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9615`

View PR using the GUI difftool: \
`$ git pr show -t 9615`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9615.diff">https://git.openjdk.org/jdk/pull/9615.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -340,8 +338,9 @@ class Patcher : public RelocActions {
     return 2;
   }
   static int adrpMovk_impl(address insn_addr, address &target) {
-    uintptr_t dest = (uintptr_t)target;
     Instruction_aarch64::patch(insn_addr + sizeof (uint32_t), 20, 5, (uintptr_t)target >> 32);
+    uintptr_t dest = (dest & 0xffffffffULL) | (uintptr_t(insn_addr) & 0xffff00000000ULL);
```
Comment (adinn) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:342]: This does not look right. `dest` is not defined on the rhs of this expression.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Err, I have no idea. Fixed.

**Review by dean-long — APPROVED** [review]
Looks good.  To detect bad addresses at emit time, you could add some asserts that check  is_valid_AArch64_address(target) in _adrp() and the patch code.  Also maybe check after patching that the desired value was rewritten using  target_addr_for_insn().

**Comment by theRealAph** [comment]
> Looks good. To detect bad addresses at emit time, you could add some asserts that check is_valid_AArch64_address(target) in _adrp() and the patch code.

I'll look at that.

 > Also maybe check after patching that the desired value was rewritten using target_addr_for_insn().

I already do that in `verify()`.

Thanks.

**Comment by theRealAph** [comment]
> > Looks good. To detect bad addresses at emit time, you could add some asserts that check is_valid_AArch64_address(target) in _adrp() and the patch code.
> 
> I'll look at that.

I'm not going to do that because I think I remember in some cases we have generated a fake "pointer" which is used as a base address. Such a value may not be a legal address. I don't know that there's anywhere we do that now, but it it is a reasonable-enough thing to do.

**Comment by theRealAph** [comment]
integrate

**Pull request closed without merging**